### PR TITLE
fix: docker guide doesn't work with prisma

### PR DESCRIPTION
**Pull request by c-ehrlich**
Closes https://github.com/t3-oss/create-t3-app/issues/389

## ✅ Checklist

- [x] I have followed every step in the [contributing guide](https://github.com/t3-oss/create-t3-app/blob/main/CONTRIBUTING.md) (updated 2022-08-15).
- [x] The PR title follows the convention we established [conventional-commit](https://www.conventionalcommits.org/en/v1.0.0/)
- [x] I performed a functional test on my final commit

---

## Changelog

- make sample environment variables more consistent (foo and bar aren't great, and they were switched between mjs files and docker instructions)
- add prisma instructions to docker guide
- didn't add a changeset as only the readme and comments were changed in cli

---

## Other notes
- Running in docker with a database in localhost requires mapping localhost to `host.docker.internal`. I added instructions for this for the the sake of completeness. But I could also see getting rid of them as I don't want anyone to think Docker is the recommended way to use CT3A in dev. What do you think?

---

## Screenshots

image

💯

### CODE REVIEW

**Comment by nexxeln** [comment]
Hey, I'm very new to docker, yesterday I was trying to deploy a t3 app to hop.io, [@alii](https://github.com/alii) helped me out and they had a very simple solution for the Dockerfile.

```Dockerfile
FROM node:16-slim
RUN apt -qy update && apt -qy install openssl
COPY package.json .
COPY package-lock.json .
COPY prisma prisma
RUN npm install
COPY . .
RUN npm run build
RUN rm .env.prod
CMD ["npm", "run", "start"]
```

Here we're installing `openssl` for prisma, copying the env variables so that the build doesn't fail and then deleting the env file so that there are no env variables in the docker image.

**Comment by c-ehrlich** [comment]
That's definitely a much simpler Dockerfile. The point of the multi-stage build is mostly to not have build artifacts in the final container, but I can also see the benefit of giving instructions for single-stage as it's way easier to read/understand, and people who need the efficiency of multi-container can implement it themselves. Not sure what's better.

Will try deploying a sample dockerized app with prisma to maybe Railway, Fly.io, and a VPS to see if there are any considerations that haven't been made yet in the current instructions - I think the most important thing about them is to be as universal as possible.

**Comment by c-ehrlich** [comment]
Sample deployment running on Railway (will delete after merge)
https://t3-docker-db-production.up.railway.app/

What does everyone think about single vs multi stage docker file? I'm leaning towards keeping it multi stage - it's only intended for prod, not development, so I think it's better to guide people towards a performant solution.

**Review by alii — APPROVED** [review]
Can't remember if `alpine` required changes to the prisma schema (another binaryTarget) but if this works then it works... It always took me ages to get prisma working in Docker esp with m1 hahah. lgtm!

**Comment by sidwebworks** [comment]
LGTM 🥇

File: cli/template/base/README.md [hunk 1]
```diff
@@ -91,9 +91,12 @@ Please note that Next.js requires a different process for buildtime (available i
    # TODO: re-evaluate if emulation is still necessary on arm64 after moving to node 18
    FROM --platform=linux/amd64 node:16-alpine AS deps
    # Check https://github.com/nodejs/docker-node/tree/b4117f9333da4138b03a546ec926ef50a31506c3#nodealpine to understand why libc6-compat might be needed.
-   RUN apk add --no-cache libc6-compat
+   RUN apk add --no-cache libc6-compat openssl
    WORKDIR /app
 
+   # Install Prisma Client - remove if not using Prisma
+   copy prisma ./
```
Comment (juliusmarminge) [inline cli/template/base/README.md:98]:
```suggestion
COPY prisma ./
```

File: cli/template/base/README.md [hunk 2]
```diff
@@ -91,9 +91,12 @@ Please note that Next.js requires a different process for buildtime (available i
    # TODO: re-evaluate if emulation is still necessary on arm64 after moving to node 18
    FROM --platform=linux/amd64 node:16-alpine AS deps
    # Check https://github.com/nodejs/docker-node/tree/b4117f9333da4138b03a546ec926ef50a31506c3#nodealpine to understand why libc6-compat might be needed.
-   RUN apk add --no-cache libc6-compat
+   RUN apk add --no-cache libc6-compat openssl
    WORKDIR /app
 
+   # Install Prisma Client - remove if not using Prisma
+COPY prisma ./
```
Comment (juliusmarminge) [inline cli/template/base/README.md:98]:
```suggestion
    COPY prisma ./
```

File: cli/template/base/README.md [hunk 3]
```diff
@@ -196,13 +201,14 @@ You can also use docker compose to build the image and run the container.
...
          args:
-           NEXT_PUBLIC_FOO: "foo"
+           NEXT_PUBLIC_CLIENTVAR: "clientvar"
        working_dir: /app
        ports:
          - "3000:3000"
        image: t3-app
        environment:
-         - BAR=bar
+         # If you are running your database on localhost in dev, replace `localhost` with `host.docker.internal`
```
Comment (juliusmarminge) [inline cli/template/base/README.md:210]: what localhost should I replace?

File: www/src/pages/en/deployment/docker.md [hunk 4]
```diff
@@ -135,10 +136,12 @@ Please note that Next.js requires a different process for buildtime (available i
 5. To build and run this image locally, run:
 
    ```bash
-   docker build -t ct3a -e NEXT_PUBLIC_FOO=foo .
-   docker run -p 3000:3000 -e BAR="bar" ct3a
+   docker build -t ct3a -e NEXT_PUBLIC_CLIENTVAR=clientvar .
+   docker run -p 3000:3000 -e DATABASE_URL="database_url_goes_here" ct3a
    ```
 
+   (If you are running your database on localhost in dev, replace `localhost` with `host.docker.internal`)
```
Comment (juliusmarminge) [inline www/src/pages/en/deployment/docker.md:143]: ^^

File: www/src/pages/en/deployment/docker.md [hunk 5]
```diff
@@ -161,13 +164,14 @@ You can also use docker compose to build the image and run the container.
...
          args:
-           NEXT_PUBLIC_FOO: "foo"
+           NEXT_PUBLIC_CLIENTVAR: "clientvar"
        working_dir: /app
        ports:
          - "3000:3000"
        image: t3-app
        environment:
-         - BAR=bar
+         # If you are running your database on localhost in dev, replace `localhost` with `host.docker.internal`
```
Comment (juliusmarminge) [inline www/src/pages/en/deployment/docker.md:173]: ^^

**Comment by juliusmarminge** [comment]
I still think its unclear that it says I should **replace localhost** yet there isn't a `localhost` to replace?

**Comment by c-ehrlich** [comment]
> I still think its unclear that it says I should **replace localhost** yet there isn't a `localhost` to replace? Is it the `database_url_goes_here` you mean?

A database running locally will have a url like `postgresql://postgres:changeme@localhost:5432/t3docker?schema=public`.

One solution to make it more obvious what is meant is to just make the sample url in that format. Another option would be to delete the text about replacing `localhost` completely - developing locally in this container is an awful experience anyway because you have to rebuild it to see changes.

**Comment by juliusmarminge** [comment]
> > I still think its unclear that it says I should **replace localhost** yet there isn't a `localhost` to replace? Is it the `database_url_goes_here` you mean?
> 
> A database running locally will have a url like `postgresql://postgres:changeme@localhost:5432/t3docker?schema=public`.
> 
> One solution to make it more obvious what is meant is to just make the sample url in that format. Another option would be to delete the text about replacing `localhost` completely - developing locally in this container is an awful experience anyway because you have to rebuild it to see changes.

Oh you mean running the app in docker in dev? Yea that we should definitely not advocate.

**Comment by c-ehrlich** [comment]
I've removed that for now.

I think it might be worth adding dev setup docker instructions (with hmr etc) at some point as the reality of many projects is that the t3 app might need to be part of a multi-container setup, even in dev. But that's better off being a separate conversation/PR in the future.

**Pull request merged**